go-controller/pkg/cni/udn/primary_network.go

```diff
@@ -0,0 +1,120 @@
+package udn
+
+import (
+	"fmt"
+
+	"k8s.io/klog/v2"
+
+	nadlister "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/listers/k8s.cni.cncf.io/v1"
+
+	cnitypes "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/cni/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+// wait on a certain pod annotation related condition
+type podAnnotWaitCond = func(map[string]string, string) (*util.PodAnnotation, bool)
+
+type UserDefinedPrimaryNetwork struct {
+	nadLister          nadlister.NetworkAttachmentDefinitionLister
+	annotation         *util.PodAnnotation
+	activeNetwork      *cnitypes.NetConf
+	findPrimaryNetwork func(namespace string, nadLister nadlister.NetworkAttachmentDefinitionLister) (*cnitypes.NetConf, error)
+}
+
+func NewPrimaryNetwork(nadLister nadlister.NetworkAttachmentDefinitionLister) *UserDefinedPrimaryNetwork {
+	return &UserDefinedPrimaryNetwork{
+		nadLister:          nadLister,
+		findPrimaryNetwork: util.FindActiveNetworkForNamespace,
+	}
+}
+
+func (p *UserDefinedPrimaryNetwork) InterfaceName() string {
+	return "sdn1"
+}
+
+func (p *UserDefinedPrimaryNetwork) Annotation() *util.PodAnnotation {
+	return p.annotation
+}
+
+func (p *UserDefinedPrimaryNetwork) NetworkName() string {
+	if p.activeNetwork == nil {
+		return ""
```

Comment:
> wouldn't this be `default` ? Also, how can we ever be in this situation ?
> 

The GetActiveNetwork logic has change and if will return an error if the active network is unknown

> Remind me: if the default network is set as the primary, it would be set, and return `default` , right ?

If the default network is the primary this do not get executed since we have this guard at the `WaitForPrimary` func
```golang
if nadName != types.DefaultNetworkName || annotation.Role == types.NetworkRolePrimary {
			return annotation, isReady
		}
```

Reply:
isn't that condition "upside down" then ? we should want to bail out early if we're on the default network, right ?

Reply:
nope, we need to bail it if:
- default network AND primary role
- non default network and non primary role

Reply:
oh of course. If it is the default network, we need to wait **iif** the default network is not primary.

Sure sure.